src/digest/sha1.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2015-2016 Brian Smith.
+// Copyright 2016 Simon Sapin.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+#![deny(unsafe_code)]
+
+use core::num::Wrapping;
+use polyfill;
+use super::{Algorithm, StateBuffer, sha256_format_output};
+
+/// SHA-1 as specified in [FIPS
+/// 180-4](http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf).
+///
+/// SHA-1 is deprecated in *ring*.
+/// The main purpose in retaining it is to support legacy protocols and OCSP,
+/// none of which need a fast SHA-1 implementation.
+/// This implementation therfore favors size and simplicity over speed.
+/// Unlike SHA-256, SHA-384, and SHA-512, there is no assembly language implementation.
+pub static SHA1: Algorithm = Algorithm {
+    output_len: 160 / 8,
+    chaining_len: 160 / 8,
+    block_len: 512 / 8,
+    len_len: 64 / 8,
+    block_data_order: sha1_block_data_order,
+    format_output: sha256_format_output,
+    initial_state: [
+        u32x2!(0x67452301u32, 0xefcdab89u32),
+        u32x2!(0x98badcfeu32, 0x10325476u32),
+        u32x2!(0xc3d2e1f0u32, 0u32),
+        0, 0, 0, 0, 0,
+    ],
+    nid: 64, // NID_sha1
+};
+
+type W32 = Wrapping<u32>;
+
+#[inline] fn ch(x: W32, y: W32, z: W32) -> W32 { (x & y) | (!x & z) }
+#[inline] fn parity(x: W32, y: W32, z: W32) -> W32 { x ^ y ^ z }
+#[inline] fn maj(x: W32, y: W32, z: W32) -> W32 { (x & y) | (x & z) | (y & z) }
+
+fn sha1_block_data_order(state: &mut StateBuffer, data: &[u8]) {
+    let state = polyfill::slice::u64_as_u32_mut(state);
+    let state = polyfill::slice::as_wrapping_mut(state);
+    let state = &mut state[..5];
+    let state = slice_as_array_ref_mut!(state, 5).unwrap();
+
+    let mut w: [W32; 80] = [Wrapping(0); 80];
+    for block in data.chunks(64) {
+        let block = slice_as_array_ref!(block, 64).unwrap();
+
+        for t in 0..16 {
+            w[t] = Wrapping(
+                (block[t * 4] as u32) << 24 |
+                (block[t * 4 + 1] as u32) << 16 |
+                (block[t * 4 + 2] as u32) << 8 |
+                (block[t * 4 + 3] as u32));
```

Comment:
1. Is this change to deal with unaligned inputs the cause of the worse performance? Or was it the conversion to `Wrapping`? Or a combination?

If the comversion to `Wrapping` caused the worse performance then we should go back to the way it was before. (And, in that case we should file a bug on the Rust stdlib.)

2. Please extract this out to a function named `from_be_u32_ptr` or similar.

3. Please use `u32::from(x)` instead of `x as u32` to follow the guidelines at https://github.com/briansmith/ring/blob/master/STYLE.md#casting-as-and-conversions.